tests/EdgeToEdge/PayPalRequestLoggerTest.php

```diff
@@ -0,0 +1,109 @@
+use WMDE\Fundraising\Frontend\App\EventHandlers\PayPalRequestLogger;
+
+/**
+ * @covers \WMDE\Fundraising\Frontend\App\EventHandlers\PayPalRequestLogger
+ */
+class PayPalRequestLoggerTest extends WebRouteTestCase {
+
+	private array $paypalRoutes = [
+		'handle_paypal_payment_notification',
+		'handle_paypal_membership_fee_payments'
+	];
+
+	private const LOG_DIR = 'log';
+	private const LOG_FILENAME = 'test.csv';
+
+	private vfsStreamDirectory $filesystem;
+	private KernelBrowser $client;
+	private string $filePath;
+
+	public function setUp(): void {
+		$this->filesystem = vfsStream::setup( self::LOG_DIR );
+		$this->filePath = vfsStream::url( self::LOG_DIR . '/' . self::LOG_FILENAME );
+
+		/** @var KernelBrowser $client */
+		$client = $this->createClient();
+		$this->client = $client;
+		self::$container->set(
+			PayPalRequestLogger::class,
+			new PayPalRequestLogger( $this->filePath, $this->paypalRoutes )
+		);
+	}
+
+	public function testWhenPostedPayPalData_createsLogFile() {
+		$route = self::$container->get( 'router' )->generate(
+			$this->paypalRoutes[0],
+			[],
+			UrlGeneratorInterface::RELATIVE_PATH
+		);
+		$payPalData = $this->validPayPalData();
+		$expectedContents = implode( ', ', $payPalData ) . PHP_EOL;
+
+		$this->client->request( 'post', $route, $payPalData );
+
+		$this->assertTrue( $this->filesystem->hasChild( self::LOG_FILENAME ) );
+		$this->assertSame( $expectedContents, file_get_contents( $this->filePath ) );
+	}
+
+	public function testWhenPostedPayPalDataAndFileExists_addsDataToLogFile() {
+		$route = self::$container->get( 'router' )->generate(
+			$this->paypalRoutes[0],
+			[],
+			UrlGeneratorInterface::RELATIVE_PATH
+		);
+		$payPalData = $this->validPayPalData();
+		$existingData = 'go, to, school, 99' . PHP_EOL;
+		$expectedContents = $existingData . implode( ', ', $payPalData ) . PHP_EOL;
+		file_put_contents( $this->filePath, $existingData );
+
+		$this->client->request( 'post', $route, $payPalData );
+
+		$this->assertSame( $expectedContents, file_get_contents( $this->filePath ) );
+	}
+
+	/**
+	 * @dataProvider routesDataProvider
+	 */
+	public function testRunsOnPayPalNotificationRoutes( string $routeName ) {
+		$route = self::$container->get( 'router' )->generate(
+			$routeName,
+			[],
+			UrlGeneratorInterface::RELATIVE_PATH
+		);
+
+		$this->client->request( 'post', $route, $this->validPayPalData() );
+
+		$this->assertTrue( $this->filesystem->hasChild( self::LOG_FILENAME ) );
+	}
+
+	public function testDoesNotRunOnNonPayPalRoute() {
+		$this->client->request( 'post', '/', $this->validPayPalData() );
+
+		$this->assertFalse( $this->filesystem->hasChild( self::LOG_FILENAME ) );
+	}
+
+	private function validPayPalData(): array {
+		return [
+			'txn_type' => 'express_checkout',
+			'txn_id' => '61E67681CH3238416',
+			'subscr_id' => '8RHHUM3W3PRH7QY6B59',
+			'item_number' => 1,
+		];
+	}
+
+	public function routesDataProvider() {
+		return [
```

Comment:
A tip (not change request) for shorter and sometimes more readbale data providers: Use `yield` instead of a two-dimensional array.